File: packages/engine/src/cores/famibob/clock.test.js
import { test, expect } from 'vitest'
import { Region, Clock, RunOption } from '.'

test('ntsc clock', () => {
  let cpu = 0
  let ppu = 0
  const clock = new Clock(
    Region.NTSC,
    () => cpu++,
    () => ppu++,
  )

  for (let i = 0; i < 60; i++) clock.run(RunOption.FRAME)

  expect(cpu / 60).toBeCloseTo(29780 + 2 / 3, 1)
  expect(ppu / 60).toBeCloseTo(89342, 1)
})

test('pal clock', () => {
  let cpu = 0
  let ppu = 0
  const clock = new Clock(
    Region.PAL,
    () => cpu++,
    () => ppu++,
  )

  for (let i = 0; i < 50; i++) clock.run(RunOption.FRAME)

  expect(cpu / 50).toBeCloseTo(33247.5, 1)
  expect(ppu / 50).toBeCloseTo(106392, 1)
})

File: packages/engine/src/cores/famibob/clock.test.zig
const std = @import("std");
const Clock = @import("clock.zig").Clock;
const Famibob = @import("famibob.zig").Famibob;

test "Famibob NTSC clock" {
    const S = struct {
        cpu: u32 = 0,
        ppu: u32 = 0,
        pub fn handleCPUCycle(self: *@This()) void {
            self.cpu += 1;
        }
        pub fn handlePPUCycle(self: *@This()) void {
            self.ppu += 1;
        }
    };

    var s = S{};
    var clock = Clock(S, .ntsc){ .handler = &s };

    for (0..60) |_| {
        clock.run(.frame);
    }

    try std.testing.expectApproxEqAbs(29780.3, @as(f64, @floatFromInt(s.cpu)) / 60.0, 0.05);
    try std.testing.expectApproxEqAbs(89341, @as(f64, @floatFromInt(s.ppu)) / 60.0, 0.05);
}

test "Famibob PAL clock" {
    const S = struct {
        cpu: u32 = 0,
        ppu: u32 = 0,
        pub fn handleCPUCycle(self: *@This()) void {
            self.cpu += 1;
        }
        pub fn handlePPUCycle(self: *@This()) void {
            self.ppu += 1;
        }
    };

    var s = S{};
    var clock = Clock(S, .pal){ .handler = &s };

    for (0..60) |_| {
        clock.run(.frame);
    }

    try std.testing.expectApproxEqAbs(33247.5, @as(f64, @floatFromInt(s.cpu)) / 60.0, 0.05);
    try std.testing.expectApproxEqAbs(106392.0, @as(f64, @floatFromInt(s.ppu)) / 60.0, 0.05);
}
